59.md

```diff
@@ -0,0 +1,55 @@
+NIP-59
+======
+
+Gift Wrap
+-------------------------------
+
+`draft` `optional` `author:kieran`
+
+This NIP defines a simple method to protect event metadata leakage by using randomly generated keys.
+
+A gift wrap event is a `kind 1059` event which wraps another event with a single use key.
+
+This event kind can be used to wrap DM's or other events which are considered private.
+
+This is similar in concept to Onion Routing.
+
+To construct a wrapped `kind 4` you first start by producing the `kind 4` event as normal, then create a random key and produce the `kind 1059` by encrypting the `kind 4` as a JSON string and storing it in `content` of the `kind 1059` using the same `p` tag to notify the recipient.
```

Comment:
```suggestion
To wrap an event, sign it as usual then serialize it to JSON. This string should then be encrypted using
NIP-44 with a different private key. Next, add the signed payload to the `content` field of a new
`kind 1059` event, signed with the same key you used to encrypt the content.

A `p` tag for the recipient MUST be added to the `tags` field on the outer event. To protect recipient identity, this may be a regular pubkey, or one representing a group of people sharing a private key for decryption.
```

Rephrased to make it clearer, and less about kind 4.

Reply:
this means if alice loses her key, all the messages she sent to bob can be read

Reply:
Do we really need all these suggestions on tags? If somebody comes to read this NIP they will be confused about why there are so many suggested tags without more context, they should come to their own realization that you can do these things, or they can be explicitly called out, its too vague imo.

Reply:
I'm not sure I understand your objection, the only specific I'm adding is `k`, which should be defined in the main table on the readme. But then you say it's too vague at the end? @earonesty and I had talked about this a bit [here](https://github.com/nostr-protocol/nips/pull/580/files#r1218625058), I added the suggestion about optionally copying tags because there are other alternatives (like encrypting tag values) that would be pretty bad. Are you looking for more examples here?

Reply:
There should not be any requirements on tags, maybe you don't even need a p tag

Reply:
None of the above are requirements, just MAY depending on whether the use case for the wrapper requires filtering (I want to build large private groups on this NIP, which would). Maybe `p` isn't strictly required, but without it wouldn't you have to attempt to decrypt all wrapped messages to find the ones you want?

Reply:
without a "p" tag, it will be impossible to find the messages you want.   but the "p" tag could, technically, be for a derived key of some kind.   for example, alice may have a public key that she uses only when talking to bob.   in that case, she would only subscribe to events of people in her contact list, and couldn't get dm's from strangers.

so yes, a "p" tag is definitely required, but it need-not be the "p" tag of the recipient.    gift-wrap is a very generic construct